### PR TITLE
Use rustup.rs instead of custom bootstrap 

### LINKED ISSUE #11361: Use rustup to get nightly rust
Since every time build servo will use nightly rust compiler, is it a good idea to use rustup and overload servo path to build servo with nightly rustc? I mean, since there is a rust version manager, why not just use it?

### LINKED ISSUE #18874: Merge ./mach bootstrap-rust and ./mach bootstrap-cargo
They share the same version numbers now anyway.

### LINKED ISSUE #19365: Doc build fails with "Unrecognized option: 'document-private-items'."
I suspect this is coming from the geckolib documentation build, which maybe uses stable rust?

```
rustdoc: Unrecognized option: 'document-private-items'.
Try 'rustdoc --help' for more information.
rustdoc: Unrecognized option: 'document-private-items'.
Try 'rustdoc --help' for more information.
rustdoc: Unrecognized option: 'document-private-items'.
Try 'rustdoc --help' for more information.
error: Could not document `log`.
```

### PR DESCRIPTION
**Pull request by SimonSapin**
Fixes #11361, closes #18874, fixes #19365.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19395)

### CODE REVIEW

**Comment by SimonSapin** [comment]
Before this can land:

* A new version of rustup that includes https://github.com/rust-lang-nursery/rustup.rs/pull/1295 needs to be released (https://github.com/rust-lang-nursery/rustup.rs/issues/1275#issuecomment-346615370)
* That version needs to be install on CI builders

At first, contributors will probably have an older version of rustup and see something like this:

```
$ ./mach build
error: Found argument '--install' which wasn't expected, or isn't valid in this context

USAGE:
    rustup run <toolchain> <command>...

For more information try --help
Build FAILED in 0:00:00
```

This is pretty bad since `--help` in the output refers to `rustup run --help`, not `./mach build --help`. And even `rustup run --help` wouldn’t be helpful since it doesn’t know about options added in later versions.

Mach cannot easily distinguish this case from a "normal" build failure, so the only way to detect this situation and print a better diagnostic we would need to run `rustup --version` and parse its output every  time we’re about to run `rustup run`. This seems overkill, but perhaps I shouldn’t worry about that.

r? @jdm

**Comment by SimonSapin** [comment]
`subprocess.check_output(["rustup", "--version"])` takes ~3ms on my laptop, so really not worth worrying about compared to existing overhead of Mach. I’ve added the detection mentioned above.

**Comment by SimonSapin** [comment]
With this PR running `cargo build --manifest-path ports/servo/Cargo.toml` will use the same compiler as `./mach build`. However going back and forth between those will still recompile everything because mach sets the `RUSTFLAGS` environment variable.

To not have `RUSTFLAGS` by default we’d need rustc to revert back to using ld.gold by default: https://github.com/rust-lang/rust/pull/29974, and us to replace `-W unused-extern-crates` with `#![warn(unused_extern_crate)]` in each crate.

**Comment by jdm** [comment]
Our Appvey and TravisCI are missing rustup; test-tidy is failing; and test-wpt is broken because ensure_bootstrapped was removed.

**Comment by SimonSapin** [comment]
Thanks. I fixed these immediate issues, but as I mentioned we’re still waiting on a new rustup release.

File: python/servo/bootstrap_commands.py
Comment (jdm) [inline python/servo/bootstrap_commands.py]: This doesn't match the `remove_anything` variable later.

File: appveyor.yml
Comment (jdm) [inline appveyor.yml]: Should we modify the cache instructions above this block?

**Review by jdm — CHANGES_REQUESTED** [review]

File: python/servo/bootstrap_commands.py
Comment (SimonSapin) [reply python/servo/bootstrap_commands.py]: This as reported by tidy and fixed.

File: appveyor.yml
Comment (SimonSapin) [reply appveyor.yml]: It looks like this caches the `.servo` directory including the extracted rustc and cargo downloaded by mach. Do we want to cache toolchains downloaded by rustup?

Comment (jdm) [reply appveyor.yml]: Yes, since it makes builds a bit faster.

Comment (SimonSapin) [reply appveyor.yml]: Alright, done.

**Comment by SimonSapin** [comment]
Rustup release is blocked on https://github.com/rust-lang-nursery/rustup.rs/issues/1305

**Comment by SimonSapin** [comment]
Added a commit to unblock by not using the new rustup feature.

@bors-servo try

**Comment by SimonSapin** [comment]
```
./mach clean-nightlies --keep 3 --force
[…]
 environment:
  PATH=/home/servo/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
[…]
OSError: [Errno 2] No such file or directory

  File "/home/servo/buildbot/slave/arm32/build/python/servo/bootstrap_commands.py", line 144, in clean_nightlies
    stdout = subprocess.check_output(['rustup', 'toolchain', 'list'])
```

`/home/servo/.cargo/bin` is not in `$PATH`. I think rustup normally adds it by writing to `~/.profile`, but maybe that file is not used in the environment where buildbot jobs run?

**Comment by SimonSapin** [comment]
This effectively runs `subprocess.call(["cargo.exe", "build"], executable="rustup.exe")` which fails on AppVeyor with `WindowsError(2, 'The system cannot find the file specified')`.

Based on reading the Python standard library, this ends up calling `CreateProcess("rustup.exe", "cargo.exe build", …)`. Per https://msdn.microsoft.com/en-us/library/windows/desktop/ms682425(v=vs.85).aspx the `PATH` environment variable is *not* use for the first parameter, only for the second.

So the only way to run `rustup.exe` on Windows with a different `argv[0]` is to do the `PATH` lookup ourselves. I don’t know if there’s a better way to do that than parsing `PATH` and looking which files exist ourselves.

**Comment by SimonSapin** [comment]
@aneeshusa, does https://github.com/servo/saltfs/pull/762 still need another step to be deployed to CI builders?

**Comment by jdm** [comment]
Yes, we need to highstate the builders.

**Comment by SimonSapin** [comment]
On AppVeyor:

```rust
error: linking with `link.exe` failed: exit code: 1120
  |
  = note: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\link.exe" "/NOLOGO" "/NXCOMPAT" "/LARGEADDRESSAWARE" "/SAFESEH" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct0.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct1.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct2.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dom_struct3.rcgu.o" "/OUT:C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dll" "/DEF:C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\rustc.Yl6vda9bXmAa\\lib.def" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.crate.metadata.rcgu.o" "C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.crate.allocator.rcgu.o" "/OPT:REF,NOICF" "/DEBUG" "/NATVIS:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\etc\\liballoc.natvis" "/NATVIS:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\etc\\libcore.natvis" "/LIBPATH:C:\\projects\\servo\\target\\debug\\deps" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "proc_macro-11829f83756c64ad.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "syntax-0f21f5e615c5859c.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "rustc_errors-1b22ef7d4be6ca6a.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "syntax_pos-7d059d588bbad9ae.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "rustc_cratesio_shim-5a0ed2b459381b6f.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "term-3095e146c74f9c19.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "rustc_data_structures-12c636f652fe56e8.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "serialize-29271e07662eeebe.dll.lib" "/LIBPATH:C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "std-14232f553fbb2f9e.dll.lib" "C:\\Users\\appveyor\\.rustup\\toolchains\\nightly-2017-11-14-i686-pc-windows-msvc\\lib\\rustlib\\i686-pc-windows-msvc\\lib\\libcompiler_builtins-c3dbf71fb0229f85.rlib" "kernel32.lib" "advapi32.lib" "ws2_32.lib" "userenv.lib" "shell32.lib" "msvcrt.lib" "/DLL" "/IMPLIB:C:\\projects\\servo\\target\\debug\\deps\\dom_struct-224c198e23df7fda.dll.lib"
  = note:    Creating library C:\projects\servo\target\debug\deps\dom_struct-224c198e23df7fda.dll.lib and object C:\projects\servo\target\debug\deps\dom_struct-224c198e23df7fda.dll.exp
          dom_struct-224c198e23df7fda.dom_struct0.rcgu.o : error LNK2019: unresolved external symbol ___CxxFrameHandler3 referenced in function ___ehhandler$_ZN4core3ptr13drop_in_place17h042bfa3a294a8463E
          dom_struct-224c198e23df7fda.dom_struct1.rcgu.o : error LNK2001: unresolved external symbol ___CxxFrameHandler3
          dom_struct-224c198e23df7fda.dom_struct2.rcgu.o : error LNK2001: unresolved external symbol ___CxxFrameHandler3
          dom_struct-224c198e23df7fda.dom_struct3.rcgu.o : error LNK2001: unresolved external symbol ___CxxFrameHandler3
          dom_struct-224c198e23df7fda.dom_struct1.rcgu.o : error LNK2019: unresolved external symbol _memcpy referenced in function __ZN33_$LT$alloc..vec..Vec$LT$T$GT$$GT$3pop17h1e5fccfad93324e0E
          dom_struct-224c198e23df7fda.dom_struct2.rcgu.o : error LNK2001: unresolved external symbol _memcpy
          dom_struct-224c198e23df7fda.dom_struct3.rcgu.o : error LNK2001: unresolved external symbol _memcpy
          dom_struct-224c198e23df7fda.dom_struct3.rcgu.o : error LNK2019: unresolved external symbol _memset referenced in function __ZN5alloc6string6String4push17h89a721a5832f9b9bE
          LINK : error LNK2001: unresolved external symbol __DllMainCRTStartup@12
          LINK : error LNK2001: unresolved external symbol __load_config_used
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\kernel32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\advapi32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\ws2_32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\userenv.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64\shell32.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64\msvcrt.lib : warning LNK4272: library machine type 'x64' conflicts with target machine type 'X86'
          C:\projects\servo\target\debug\deps\dom_struct-224c198e23df7fda.dll : fatal error LNK1120: 5 unresolved externals
          
error: aborting due to previous error
error: Could not compile `dom_struct`.

Caused by:
  process didn't exit successfully: `rustc --crate-name dom_struct components\dom_struct\lib.rs --crate-type proc-macro --emit=dep-info,link -C prefer-dynamic -C codegen-units=4 -C debuginfo=2 -C metadata=224c198e23df7fda -C extra-filename=-224c198e23df7fda --out-dir C:\projects\servo\target\debug\deps -L dependency=C:\projects\servo\target\debug\deps -W unused-extern-crates` (exit code: 101)

**Comment by SimonSapin** [comment]
Remaining buildbot failures are:

* On OS X `~/.cargo/bin` is not in `$PATH` (https://github.com/servo/saltfs/pull/762). Additionally, as far as can tell rustup in installed at that location on some builders but not others (https://github.com/servo/saltfs/pull/758/files and https://github.com/servo/saltfs/pull/767).
* On Windows `C:\Users\Administrator\.cargo\bin` is in `$PATH` but rustup is apparently not installed.

**Comment by SimonSapin** [comment]
On AppVeyor:

```
mach build -d -v && mach test-unit
info: syncing channel updates for 'nightly-2017-11-14-i686-pc-windows-msvc'
```
and then:
```
 = note: "C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\amd64\\link.exe"  […]
```

So it looks like we’re using a 64-bit linker with a 32-bit compiler.

**Comment by SimonSapin** [comment]
Ok, so now `$PATH` looks correct:

```
PATH=/home/servo/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

But calling rustup from mach still doesn’t work:

```python

OSError: [Errno 2] No such file or directory

  File "/Users/servo/buildbot/slave/mac-rel-wpt4/build/python/servo/bootstrap_commands.py", line 144, in clean_nightlies
    stdout = subprocess.check_output(['rustup', 'toolchain', 'list'])
```

It does work when I run the same Python code via SSH, though:

```sh
% ssh root@servo-mac2.servo.org
servo-mac2:~ root# su - servo
servo-mac2:~ servo$ echo $PATH
/Users/servo/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
servo-mac2:~ servo$ which rustup
/Users/servo/.cargo/bin/rustup
servo-mac2:~ servo$ rustup -V
rustup 1.8.0 (cb2d14862 2017-12-17)
servo-mac2:~ servo$ python
>>> import subprocess
>>> subprocess.check_output(['rustup', 'toolchain', 'list'])
'no installed toolchains\n'
```

**Comment by SimonSapin** [comment]
Ah, it should be `/Users` in `PATH` instead of `/home`. Looks like `{{ common.servo_home }}` in https://github.com/servo/saltfs/pull/771/files is wrong…

**Comment by SimonSapin** [comment]
`common/map.jinja`

```python
{%
  set common = salt['grains.filter_by']({
      'defaults': {
        'servo_home': '/home/servo',
      },
      'MacOS': {
        'servo_home': '/Users/servo'
      }
    },
    base='defaults',
    grain='os'
  )
%}
```

@aneeshusa, does this look right?

**Comment by SimonSapin** [comment]
https://github.com/servo/saltfs/pull/758#discussion_r160287667

**Comment by SimonSapin** [comment]
All non-Windows jobs are now green \o/

**Comment by SimonSapin** [comment]
@KiChjang This is https://github.com/servo/saltfs/issues/773.

File: python/servo/post_build_commands.py
Comment (jdm) [inline python/servo/post_build_commands.py]: My only question is whether `./mach cargo-gecko doc` behaves as expected with these changes, since we use it in upload_docs.sh.

**Review by jdm — APPROVED** [review]
My only question is whether `./mach cargo-gecko doc` behaves as expected with these changes, since we use it in upload_docs.sh.

**Comment by jdm** [comment]
That test failure is #19033.

**Comment by SimonSapin** [comment]
Hmm, I get "rustdoc: Unrecognized option: 'document-private-items'" errors with `mach cargo-geckolib doc` but I wonder why they didn’t occur on master. Anyway, reverting d96f0ff6a77e309682898cc7e1905a7bd0feb3cf (until we upgrade the stable toolchain to a recent enough version) fixes it.

**Comment by jdm** [comment]
Oh yeah, I forgot that https://github.com/servo/servo/issues/19365 has existed for multiple months now :(

**Comment by SimonSapin** [comment]
```
KeyError: u'RUSTDOCFLAGS'

  File "/Users/servo/buildbot/slave/mac-dev-unit/build/python/servo/testing_commands.py", line 259, in test_unit
    del env["RUSTDOCFLAGS"]
```

Replaced `del` with `pop()`.

@bors-servo r=jdm

**Comment by jdm** [comment]
```
   Doc-tests script_tests
error: An unknown error occurred

To learn more, run the command again with --verbose.
```

File: python/servo/command_base.py
Comment (jdm) [inline python/servo/command_base.py]: Given the appveyor failure (`thread 'main' panicked at 'OpenSSL library directory does not exist: C:\projects\servo\.servo\msvc-dependencies\openssl\1.1.0e-vs2015\lib64', C:\Users\appveyor\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-sys-0.9.22\build.rs:69:8`), I fear we may have lost something important by removing this line.

Comment (SimonSapin) [reply python/servo/command_base.py]: Reverted.

**Comment by SimonSapin** [comment]
Ugh, the `RUSTDOCFLAGS` revert causes warnings in `./mach test-unit` for every crate that has doctests enabled:

```
WARNING: the 'no-defaults' flag is considered deprecated
WARNING: please see https://github.com/rust-lang/rust/issues/44136
WARNING: the 'passes' flag is considered deprecated
WARNING: please see https://github.com/rust-lang/rust/issues/44136
WARNING: (you may want to use --document-private-items)
```

**Review by jdm — APPROVED** [review]

**Pull request merged**
